torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -0,0 +1,624 @@
+            )
+        orig_dtype = next(iter(orig_dtypes))
+        reduce_dtypes = {fsdp_param.reduce_dtype for fsdp_param in self.fsdp_params}
+        if len(reduce_dtypes) != 1:
+            # This can be relaxed if we issue one reduce-scatter per reduce
+            # dtype (but we would need a way for users to specify multiple
+            # reduce dtypes)
+            raise AssertionError(
+                f"FSDP expects uniform reduce dtype but got {reduce_dtypes}"
+            )
+        reduce_dtype = next(iter(reduce_dtypes))
+        return orig_dtype, reduce_dtype
+
+    # Runtime #
+    def _unshard(self, async_op: bool = False):
+        if self._all_gather_result is not None:  # manually unsharded
+            return
+        log.info("_unshard for %s", self._module_fqn)
+        if self._reshard_after_forward_event is not None:
+            # If previously resharded to a different world size, the resharded
+            # parameter data are allocated in the default stream and used in
+            # the all-gather streams.
+            self.all_gather_copy_in_stream.wait_event(self._reshard_after_forward_event)
+            self.all_gather_stream.wait_event(self._reshard_after_forward_event)
+            self._reshard_after_forward_event = None
+        self._all_gather_result = foreach_all_gather(
+            self.fsdp_params,
+            self._all_gather_process_group,
+            async_op,
+            self._all_gather_copy_in_stream_for_unshard,
+            self._all_gather_stream_for_unshard,
+            self._use_uint8_all_gather,
+            self._device,
+        )
+
+    def _wait_for_unshard(self):
+        """
+        1. In forward with implict prefetching, to overlap the current copy-out
+        with the next all-gather, we save a reference to the current all-gather
+        result to free after the next copy-out.
+        2. Otherwise (explicit prefetching or in backward), we free the
+        all-gather result immediately after the current copy-out.
+        """
+        if not self._all_gather_result:
+            return  # no preceding `_unshard()`
+        if (event := self._all_gather_result.all_gather_event) is not None:  # sync
+            torch.cuda.current_stream().wait_event(event)
+        if (work := self._all_gather_result.all_gather_work) is not None:  # async
+            work.wait()
+        if self._training_state == TrainingState.FORWARD:  # implicit prefetch
+            if prev_all_gather_state := self.all_gather_state.pop():
+                self._wait_all_gather_streams_on_event(prev_all_gather_state.event)
+                del prev_all_gather_state  # free
+        foreach_all_gather_copy_out(
+            self._all_gather_result.all_gather_output,
+            self.fsdp_params,
+            self._all_gather_process_group,
+            self._use_uint8_all_gather,
+        )
+        for fsdp_param in self.fsdp_params:
+            fsdp_param.to_unsharded()
```

Comment:
This is where the all-gather copy-out gets called.

By transitioning the FSDP parameters via `to_unsharded()`, we use `setattr(module, param_name, unsharded_param)` to give the unsharded parameter to the module for forward computation.